src/Runner.Worker/ExecutionContext.cs

```diff
@@ -266,6 +267,64 @@ public void RegisterPostJobStep(IStep step)
             Root.PostJobSteps.Push(step);
         }
 
+        /*
+            RegisterCompositeStep is a helper function used in CompositeActionHandler::RunAsync to 
+            add a child node, aka a step, to the current job to the front of the queue for processing. 
+        */
+        public IStep RegisterCompositeStep(IStep step, DictionaryContextData inputsData)
+        {
+            // ~Brute Force Method~
+            // There is no way to put this current job in front of the queue in < O(n) time where n = # of elements in JobSteps
+            // Everytime we add a new step, you could requeue every item to put those steps from that stack in JobSteps which 
+            // would result in O(n) for each time we add a composite action step where n = number of jobSteps which would compound 
+            // to O(n*m) where m = number of composite steps
+            // var temp = Root.JobSteps.ToArray();
+            // Root.JobSteps.Clear();
+            // Root.JobSteps.Enqueue(step);
+            // foreach(var s in temp)
+            //     Root.JobSteps.Enqueue(s);
+
+            // ~Optimized Method~
+            // Alterative solution: We add to another temp Queue
+            // After we add all the transformed composite steps to this temp queue, we requeue the whole JobSteps accordingly in EnqueueAllCompositeSteps()
+            // where the queued composite steps are at the front of the JobSteps Queue and the rest of the jobs maintain its order and are 
+            // placed after the queued composite steps
+            // This will take only O(n+m) time which would be just as efficient if we refactored the code of JobSteps to a PriorityQueue
+            // This temp Queue is created in the CompositeActionHandler. 
+            var newGuid = Guid.NewGuid();
```

Comment:
I'm worried a little bit because we're not adding the contexts from the job message, like we normally do when creating steps.

I know this is what Post does, but wouldnt surprise me if there is a bug there too.

It may be safe, since we're copying the root node (maybe job message contexts copied there first so safe?)

Let's add a TODO comment in the code like "// TODO: confirm whether not copying message contexts is safe".

I can help look into this more. It's tedious because values are delay expanded sometimes (sometimes evaluated early, like DisplayName). I wish the code were designed simpler... hmm thinking about alternatives.

Reply:
Added TODO. Leaving open for visibility.

Reply:
After more testing for the env flow, I realized that the display name does not evaluate the env variables. This has to do with not passing the context. See: https://github.com/actions/runner/pull/557#issuecomment-647539455

Building in support will occur in the env flow PR: https://github.com/actions/runner/pull/557

Reply:
There are probably more reasons that we need context but that is just one reason ^

Reply:
In the env flow PR, I handle the contexts for each correctly: https://github.com/actions/runner/pull/557/files#diff-e3d54b1d0cbdce4d358272df29857cf2R225 by setting the step.EnvironmentVariables

Demo: https://github.com/ethanchewy/testing-actions/actions/runs/143774560

Reply:
Also, since we add each of our composite steps to JobSteps, eventually each of the steps will be run via the ScriptHandler. 

In the ScriptHandler, it exposes the context to the step/job: https://github.com/actions/runner/blob/master/src/Runner.Worker/Handlers/ScriptHandler.cs#L248-L258